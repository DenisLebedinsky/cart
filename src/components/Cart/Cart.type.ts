import { addProduct, saveCart } from "../../redux/reducers/cart/actions";
import { Product } from "../../redux/reducers/cart/types";

export interface CartProps {
  products: Product[];
  addProduct: typeof addProduct;
  saveCart: typeof saveCart;
}
