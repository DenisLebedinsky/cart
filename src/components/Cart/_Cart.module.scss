@import "../../sass/mixins";

.cart {
  text-align: center;
  padding: 40px;

  h1 {
    margin-bottom: 40px;
  }

  > div:first-of-type {
    display: flex;
    justify-content: space-around;
    margin-bottom: 40px;

    > button {
      @include white-button;
    }
  }

  > h2 {
    font-size: 24px;
    font-weight: bold;
    text-transform: uppercase;
  }
}
