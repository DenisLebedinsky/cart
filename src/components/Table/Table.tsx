import * as React from "react";

import TableRow from "../../containers/TableRowContainer";
import TableHeaderCell from "../TableHeaderCell";
import { stableSort, getSorting } from "../../utils/sort";

import css from "./_Table.module.scss";

import { Product } from "../../redux/reducers/cart/types";
import { orderBy, TableProps, TableState } from "./Table.type";

class Table extends React.Component<TableProps, TableState> {
  public state: Readonly<TableState> = {
    order: "DESC",
    orderBy: null
  };

  public handleClick = (column: orderBy) => () => {
    let { order } = this.state;

    if (column === this.state.orderBy && order === "DESC") {
      order = "ASC";
    } else {
      order = "DESC";
    }

    this.setState({ order, orderBy: column });
  };

  public getClassName = (column: orderBy) => {
    const { orderBy, order } = this.state;

    if (orderBy === column) {
      return order === "DESC" ? css.desc : css.asc;
    }

    return "";
  };

  public renderItems = () => {
    const { orderBy, order } = this.state;
    let { items } = this.props;

    if (orderBy) {
      items = stableSort(items, getSorting(order, orderBy));
    }

    return items.map((item: Product) => (
      <TableRow
        key={item.ProductId}
        id={item.ProductId}
        name={item.Name}
        quantity={item.Quantity}
        price={item.Price}
      />
    ));
  };

  public render() {
    return (
      <div className={css.table}>
        <div className={css.header}>
          <TableHeaderCell
            columnName="Product"
            className={this.getClassName("Name")}
            onClick={this.handleClick("Name")}
          />
          <TableHeaderCell
            columnName="Quantity"
            className={this.getClassName("Quantity")}
            onClick={this.handleClick("Quantity")}
          />
          <TableHeaderCell
            columnName="Price"
            className={this.getClassName("Price")}
            onClick={this.handleClick("Price")}
          />
          <TableHeaderCell />
        </div>
        {this.renderItems()}
      </div>
    );
  }
}

export default Table;
