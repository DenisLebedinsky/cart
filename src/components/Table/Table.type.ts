import { Product } from "../../redux/reducers/cart/types";

export interface TableState {
  orderBy: orderBy;
  order: order;
}

export interface TableProps {
  items: Product[];
}

export type order = "ASC" | "DESC";

export type orderBy = keyof PartialProduct | null;

type PartialProduct = Pick<Product, "Name" | "Quantity" | "Price">;
