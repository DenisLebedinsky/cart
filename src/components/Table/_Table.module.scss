.table {
  margin: 20px auto;
  background-color: #F5F5F5;
  min-width: 510px;
  max-width: 1200px;
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.4);

  .header {
    display: flex;

    > div {
      padding: 10px;

      > p {
        text-align: center;
        font-size: 18px;
        font-weight: bold;
        cursor: pointer;
      }

      &:first-child {
        flex: 4;
      }

      &:nth-child(2) {
        flex: 3;
      }

      &:nth-child(3) {
        flex: 3;
      }

      &:last-child {
        flex: 0 0 auto;
        width: 120px;
      }

      &.asc {
        display: flex;
        justify-content: center;
        align-items: center;

        &:after {
          content: "▲";
          margin-left: 10px;
        }
      }

      &.desc {
        display: flex;
        justify-content: center;
        align-items: center;

        &:after {
          content: "▼";
          margin-left: 10px;
        }
      }
    }
  }

  > div {
    border-bottom: 1px solid #000000;
    border-left: 1px solid #000000;
    border-right: 1px solid #000000;

    &:first-child {
      border-top: 1px solid #000000;
    }
  }
}
