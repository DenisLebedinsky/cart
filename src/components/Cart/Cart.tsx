import * as React from "react";

import Header from "../Header";
import Table from "../Table";

import css from "./_Cart.module.scss";

import { CartProps } from "./Cart.type";

export class Cart extends React.Component<CartProps> {
  public renderProducts = () => {
    const { products } = this.props;

    if (products.length) {
      return <Table items={products} />;
    }

    return (
      <>
        <h2>The cart is empty</h2>
        <h2>Please click "ADD PRODUCT" button to add an item</h2>
      </>
    );
  };

  public render() {
    return (
      <div className={css.cart}>
        <Header title="Cart" />
        <div>
          <button onClick={this.props.addProduct}>Add Product</button>
          <button onClick={this.props.saveCart}>Save Cart</button>
        </div>
        {this.renderProducts()}
      </div>
    );
  }
}

export default Cart;
