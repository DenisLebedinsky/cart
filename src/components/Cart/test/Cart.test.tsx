import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import Cart from "../Cart";
import Table from "../../Table";

describe("Cart", () => {
  let wrapper: ShallowWrapper<Cart>;
  let products: any[] = ["marlar", "foobar"];

  let addProduct = jest.fn();
  let saveCart = jest.fn();

  beforeEach(() => {
    addProduct = jest.fn();
    saveCart = jest.fn();

    wrapper = shallow(
      <Cart products={products} addProduct={addProduct} saveCart={saveCart} />
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render <Table /> when products is not empty", () => {
    expect(wrapper.contains(<Table items={products}/>)).toBe(true);
    expect(wrapper.find(Table)).toHaveLength(1);
  });

  it("should call render two 'h2's when products is empty", () => {
    products = [];
    wrapper = shallow(
      <Cart products={products} addProduct={addProduct} saveCart={saveCart} />
    );

    expect(wrapper.find("h2").length).toBe(2);
  });

  it("should call addProduct with products' length", () => {
    products = [];
    wrapper = shallow(
      <Cart products={products} addProduct={addProduct} saveCart={saveCart} />
    );

    expect(wrapper.find("h2").length).toBe(2);
  });
});
