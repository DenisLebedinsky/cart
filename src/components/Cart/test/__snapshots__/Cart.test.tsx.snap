// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should match the snapshot 1`] = `
<div
  className="cart"
>
  <Header
    title="Cart"
  />
  <div>
    <button
      onClick={[MockFunction]}
    >
      Add Product
    </button>
    <button
      onClick={[MockFunction]}
    >
      Save Cart
    </button>
  </div>
  <Table
    items={
      Array [
        "marlar",
        "foobar",
      ]
    }
  />
</div>
`;
